src/differentials.jl

```diff
@@ -206,3 +206,27 @@ end
 end
 
 Base.conj(x::Thunk) = @thunk(conj(extern(x)))
+
+
+####
+"""
+    NO_FIELDS
+
+Constant for the reverse-mode derivative with respect to a structure that has no fields.
+The most notable use for this is for the reverse-mode derivative with respect to the
+function itself, when that function is not a closure.
+"""
+const NO_FIELDS = DNE()
+
+####
+"""
+    differential(𝒟::Type, der)
+
```

Comment:
I'm a little confused by what role this plays in the API

I see it's use in `[wirtinger_]propagation_expr`, but when would it be used otherwise?

Reply:
At present it is only used in 
`wirtinger_propagation_expr` it does the things 
where if you know you are in a real domain (because the input point was real)
and you know the wirtinger derviative, then convert that to a real derviative.

I can imagine us doing other things with it later.
Some speculations:
 - We might accept other forms of generalized derivatives (What is Wirtinger for Quaternions?)
 - We might like to do `differential(::Type{<:Number}, der::Number) = iszero(der) ? Zero() : der`
 
But that is out of scope of this PR.